amlip_cpp/src/cpp/types/GenericType.cpp

```diff
@@ -0,0 +1,214 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/*!
+ * @file GenericType.cpp
+ * This source file contains the definition of a generic type that contains void* data.
+ */
+
+#ifdef _WINID_SIZE
+// Remove linker warning LNK4221 on Visual Studio
+namespace {
+char dummy;
+}  // namespace
+#endif  // _WINID_SIZE
+
+#include <fastcdr/Cdr.h>
+
+#include <fastcdr/exceptions/BadParamException.h>
+using namespace eprosima::fastcdr::exception;
+
+#include <array>
+#include <stdlib.h>
+
+#include <types/GenericType.hpp>
+
+namespace eprosima {
+namespace amlip {
+namespace types {
+
+GenericType::GenericType(
+        void* data,
+        const uint32_t size)
+    : data_(data)
+    , data_size_(size)
+    , has_been_allocated_(false)
+{
+}
+
+GenericType::GenericType()
+    : GenericType(nullptr, 0)
+{
+}
+
+GenericType::~GenericType()
+{
+    // In case the data has been allocated from this class, we free it.
+    if (has_been_allocated_)
+    {
+        free(data_);
+    }
+}
+
+GenericType::GenericType(
+        const GenericType& x)
+{
+    data_ = x.data_;
+    data_size_ = x.data_size_;
+}
+
+GenericType::GenericType(
+        GenericType&& x)
+{
+    data_ = std::move(x.data_);
+    data_size_ = std::move(x.data_size_);
+}
+
+GenericType& GenericType::operator =(
+        const GenericType& x)
+{
+    data_ = x.data_;
+    data_size_ = x.data_size_;
+
+    return *this;
+}
+
+GenericType& GenericType::operator =(
+        GenericType&& x)
+{
+    data_ = std::move(x.data_);
+    data_size_ = x.data_size_;
+
+    return *this;
+}
+
+bool GenericType::operator ==(
+        const GenericType& x) const
+{
+    return (data_ == x.data_ && data_size_ == x.data_size_);
+}
+
+bool GenericType::operator !=(
+        const GenericType& x) const
+{
+    return !(*this == x);
+}
+
+void* GenericType::data() const
+{
+    return data_;
+}
+
+uint32_t GenericType::data_size() const
+{
+    return data_size_;
+}
+
+const char* GenericType::type_name()
+{
+    return "GENERIC";
+}
+
+void GenericType::serialize(
+        eprosima::fastcdr::Cdr& scdr) const
+{
+    scdr << data_size_;
+    scdr.serializeArray(static_cast<uint8_t*>(data_), data_size_);
+}
+
+void GenericType::deserialize(
+        eprosima::fastcdr::Cdr& dcdr)
+{
+    // If data has been already allocated (it has been already deserialized), we free it
+    if (has_been_allocated_)
+    {
+        free(data_);
+    }
+
+    dcdr >> data_size_;
+
+    // Store enough space to deserialize the data
+    data_ = malloc(data_size_);
```

Comment:
NIT:
It is not needed, but maybe set it explicitly would be easier to read.

```suggestion
    data_ = malloc(data_size_ * size_of(uint32_t));
```

Reply:
I guess you meant `uint8_t`